Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+    }
+
+all_headers_read:
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Could not find primary volume");
+        return false;
+    }
+
+    m_logical_block_size = LittleEndian { m_primary_volume->logical_block_size.little };
+    return true;
+}
+
+bool ISO9660FS::create_root_inode()
+{
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Primary volume doesn't exist, can't create root inode");
+        return false;
+    }
+
+    m_root_inode = ISO9660Inode::create_from_directory_record(*this, &m_primary_volume->root_directory_record_header, {});
+    return true;
+}
+
+bool ISO9660FS::calculate_inode_count()
+{
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Primary volume doesn't exist, can't calculate inode count");
+        return false;
+    }
+
+    m_cached_inode_count = 1;
+
+    auto traversal_result = visit_directory_record(&m_primary_volume->root_directory_record_header, [&](ISO::DirectoryRecordHeader const* header) {
+        if (header == nullptr) {
+            return RecursionDecision::Continue;
+        }
+
+        m_cached_inode_count += 1;
+
+        if (has_flag(header->file_flags, ISO::FileFlags::Directory)) {
+            if (header->file_identifier_length == 1) {
+                auto file_identifier = reinterpret_cast<u8 const*>(header + 1);
+                if (file_identifier[0] == '\0' || file_identifier[0] == '\1') {
+                    return RecursionDecision::Continue;
+                }
+            }
+
+            return RecursionDecision::Recurse;
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error()) {
+        dbgln_if(ISO9660_DEBUG, "Failed to traverse for caching inode count: {}", traversal_result);
+        return false;
+    }
+
+    return true;
+}
+
+KResult ISO9660FS::visit_directory_record(ISO::DirectoryRecordHeader const* record, Function<KResultOr<RecursionDecision>(ISO::DirectoryRecordHeader const*)> const& visitor) const
+{
+    if (!has_flag(record->file_flags, ISO::FileFlags::Directory)) {
+        return KSuccess;
+    }
+
+    ISO9660DirectoryIterator iterator { const_cast<ISO9660FS&>(*this), record };
+
+    while (!iterator.done()) {
+        auto maybe_result = visitor(*iterator);
+        if (maybe_result.is_error()) {
+            return maybe_result.error();
+        }
+
+        auto result = maybe_result.value();
+        if (result == RecursionDecision::Recurse) {
+            auto maybe_has_moved = iterator.next();
+            if (maybe_has_moved.is_error()) {
+                return maybe_has_moved.error();
+            }
+
+            if (!maybe_has_moved.value()) {
+                // If next() hasn't moved then we have read through all the
+                // directories, and can exit.
+                return KSuccess;
+            }
+
+            continue;
+        } else if (result == RecursionDecision::Break) {
+            return KSuccess;
+        } else {
```

Comment:
remember the "no `else` after `return`" thing?
guess what, it applies here too!

Reply:
It tells you how each `RecursionDecision` value is handled more explicitly. I could use a `switch` instead.

Reply:
Not having it also tells you that.

if (result == one thing) {
    ....
    continue;
}
if (result == other thing) {
    return ...;
}
// Otherwise
 more stuff
```

Reply:
That really does feel like an "implicit default" thing which I don't enjoy when handling enumeration values. I'll switch to a `switch`.